llvm/docs/DirectX/DXILResources.rst

```diff
@@ -0,0 +1,799 @@
+
+The "dx.Sampler" type is used to represent sampler state. The sampler type is
+an enum value from the DXIL ABI, and these appear in sampling operations as
+well as LOD calculations and texture gather.
+
+Constant Buffers
+----------------
+
+.. code-block:: llvm
+
+   target("dx.CBuffer", BufferSize)
+
+The "dx.CBuffer" type is a constant buffer of the given size. Note that despite
+the name this is distinct from the buffer types, and can only be read using the
+"dx.CBufferLoad" and "dx.CBufferLoadLegacy" operations.
+
+Buffers
+-------
+
+.. code-block:: llvm
+
+   target("dx.Buffer", ElementType, IsWriteable, IsROV)
+
+There is only one buffer type. This can represent both UAVs and SRVs via the
+``IsWriteable`` field. Since the type that's encoded is an llvm type, it
+handles both ``Buffer`` and ``StructuredBuffer`` uniformly. For ``RawBuffer``,
+the type is ``i8``, which is unambiguous since ``char`` isn't a legal type in
+HLSL.
+
+These types are generally used by BufferLoad and BufferStore operations, as
+well as atomics.
+
+There are a few fields to describe variants of all of these types:
+
+.. list-table:: Buffer Fields
+   :header-rows: 1
+
+   * - Field
+     - Description
+   * - ElementType
+     - Type for a single element, such as ``i8``, ``v4f32``, or a structure
+       type.
+   * - IsWriteable
+     - Whether or not the field is writeable. This distinguishes SRVs (not
+       writeable) and UAVs (writeable).
+   * - IsROV
+     - Whether the UAV is a rasterizer ordered view. Always ``0`` for SRVs.
+
+Textures
+--------
+
+.. code-block:: llvm
+
+   target("dx.Texture1D", ElementType, IsWriteable, IsROV)
+   target("dx.Texture1D", ...)
+   target("dx.Texture1DArray", ...)
+   target("dx.Texture2D", ...)
+   target("dx.Texture2DArray", ...)
+   target("dx.Texture3D", ...)
+   target("dx.TextureCUBE", ...)
+   target("dx.TextureCUBEArray", ...)
+
+   target("dx.Texture2DMS", ElementType, IsWriteable, IsROV, SampleCount)
+   target("dx.Texture2DMSArray", ...)
+
+   target("dx.FeedbackTexture2D", ElementType, IsWriteable, IsROV, FeedbackType)
+   target("dx.FeedbackTexture2DArray", ...)
+
+There are a number of texture types, but they are mostly interestingly
+different in their dimensions. These are distinct so that we can overload the
+various sample and texture load/store operations such that their parameters are
+appropriate to the type.
+
+.. list-table:: Buffer Fields
+   :header-rows: 1
+
+   * - Field
+     - Description
+   * - ElementType
+     - Type for a single element, such as ``i8``, ``v4f32``, or a structure
+       type.
+   * - IsWriteable
+     - Whether or not the field is writeable. This distinguishes SRVs (not
+       writeable) and UAVs (writeable).
+   * - SampleCount
+     - Sample count for a multisampled texture.
+   * - FeedbackType
+     - Feedback type for a feedback texture.
+
+.. note:: TODO: Should dimensions be specified as a type?
+
+   We could probably get rid of the 1D/2D/etc distinctions...
```

Comment:
> Should dimensions be specified as a type?

Seems that you made a good argument earlier for why not to do this above?

> so that we can overload the various sample and texture load/store operations such that their parameters are appropriate to the type.

Reply:
The TODO under "Texture Load and Store" covers what I was getting at here more thoroughly. I've removed this one.